cmd/server/main.go

```diff
@@ -58,109 +81,105 @@ func main() {
 		logger.SetLevel(level)
 	}
 
-	// create new tcp listenerf
-	ln, err := net.Listen("tcp", ServerAddress)
-	if err != nil {
-		logger.Fatalln(err)
-	}
-
-	interceptors := []grpc.UnaryServerInterceptor{
-		// validation interceptor
-		grpc_validator.UnaryServerInterceptor(),
-		// logger interceptor
-		grpc_logrus.UnaryServerInterceptor(logrus.NewEntry(logger)),
-		// Request-Id interceptor
-		requestid.UnaryServerInterceptor(),
-	}
-	// add authorization interceptor if authz service address is provided
-	if AuthzAddr != "" {
-		interceptors = append(interceptors,
-			// authorization interceptor
-			toolkit_auth.UnaryServerInterceptor(AuthzAddr, applicationID),
-		)
-	}
-	middleware := grpc_middleware.ChainUnaryServer(interceptors...)
-	// create new gRPC server with middleware chain
-	server := grpc.NewServer(
-		grpc.UnaryInterceptor(
-			middleware,
-		),
-	)
+	return logger
+}
 
+// ServeInternal builds and runs the server that listens on InternalAddress
+func ServeInternal(logger *logrus.Logger) error {
 	healthChecker := health.NewChecksHandler("/healthz", "/ready")
 	healthChecker.AddReadiness("DB ready check", dbReady)
-	go http.ListenAndServe(HealthAddress, healthChecker.Handler())
-
-	// waiting for database is available.
-	logger.Info("Connecting to database...")
-	dbCheckContext, cancel := context.WithTimeout(context.Background(), time.Minute*5)
-	for {
-		select {
-		case <-time.After(time.Second * 3):
-			if err = dbReady(); err != nil {
-				continue
-			}
-			cancel()
-		case <-dbCheckContext.Done():
-			cancel()
-		}
-		break
+	healthChecker.AddLiveness("ping", health.HTTPGetCheck(fmt.Sprint("http://", InternalAddress, "/ping"), time.Minute))
+
+	s, err := server.NewServer(
+		// register our health checks
+		server.WithHealthChecks(healthChecker),
+		// this endpoint will be used for our health checks
+		server.WithHandler("/ping", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(200)
+			w.Write([]byte("pong"))
+		})),
+	)
+	if err != nil {
+		return err
 	}
+	l, err := net.Listen("tcp", InternalAddress)
 	if err != nil {
-		logger.Fatalln(err)
+		return err
 	}
-	logger.Info("Connected to database successfully.")
+
+	logger.Debugf("serving internal http at %q", InternalAddress)
+
+	return s.Serve(nil, l)
+}
+
+// ServeExternal builds and runs the server that listens on ServerAddress and GatewayAddress
+func ServeExternal(logger *logrus.Logger) error {
 	dbSQL, err := sql.Open("postgres", DBConnectionString)
 	if err != nil {
 		logger.Fatal(err)
```

Comment:
I'm wondering if `logger.Fatal(err)` is the right choice here (same goes for line 122) because the other error checks in this function simply return the `err` value. Is there a reason for this? Just bringing it up for the sake of consistency.

I also think `fatal` will stop deferred functions from executing, which (I think) would keep an unused database connection alive.